### PR TITLE
New features

### PR DESCRIPTION
**Pull request by amartiushkov-ms**
- Added reporting of Before/After hooks;
- Added ability to rerun launch;
- Added retry option;
- Added is_skipped_an_issue option;
- Updated env var name for skipping of GA.

### CODE REVIEW

File: behave_reportportal/behave_agent.py [hunk 1]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
             self._log_step_exception(step, self._scenario_id)
 
     def _log_step_exception(self, step, item_id):
+        message = [
+            "Step [{keyword}]: {name} was finished with exception.".format(
+                keyword=step.keyword, name=step.name
+            )
+        ]
         if step.exception:
+            if step.exception:
```
Comment (HardNorth) [inline behave_reportportal/behave_agent.py]: The statement is always true

File: behave_reportportal/behave_agent.py [hunk 2]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+                    name=msg,
+                    start_time=timestamp(),
+                    item_type="AFTER_TEST",
+                    parent_item_id=self._scenario_id,
+                )
+                self._rp.finish_test_item(self._step_id, timestamp(), "PASSED")
+            else:
+                self._rp.log(
+                    timestamp(), msg, level="INFO", item_id=self._scenario_id
+                )
```
Comment (HardNorth) [inline behave_reportportal/behave_agent.py]: Two methods contain almost identical code. That needs refactoring.

**Review by HardNorth — CHANGES_REQUESTED** [review]

File: README.rst [hunk 3]
```diff
@@ -59,9 +60,14 @@ The following parameters are optional:
 
 - :code:`launch_name = AnyLaunchName` - launch name (default value is 'Python Behave Launch')
 - :code:`launch_id = xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` - id of the existing launch (the session will not handle the lifecycle of the given launch)
-- :code:`launch_attributes = 'Smoke Env:Python3'` - list of attributes for launch
-- :code:`launch_description = 'Smoke test'` - launch description
+- :code:`launch_attributes = Smoke Env:Python3` - list of attributes for launch
+- :code:`launch_description = Smoke test` - launch description
 - :code:`step_based = True` - responsible for Scenario or Step based logging (Scenario based approach is used by default)
+- :code:`is_skipped_an_issue = False` - if it True, skipped tests are not marked as issue
```
Comment (iivanou) [inline README.rst]:
```suggestion
- :code:`is_skipped_an_issue = False` - a flag to skip tests which are not marked as an issue
```

File: behave_reportportal/behave_agent.py [hunk 4]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+            time=timestamp(),
+            level="ERROR",
+            message="\n".join(message),
+        )
+
+    def _log_fixtures(self, item, item_type, parent_item_id):
+        """
+        Log used fixtures for item.
+
+        It will be log records for scenario based approach
```
Comment (iivanou) [inline behave_reportportal/behave_agent.py]: `It will be logging` or `It will log`

File: behave_reportportal/behave_agent.py [hunk 5]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+        and step for step based.
+        """
+        if not item.tags:
```
Comment (iivanou) [inline behave_reportportal/behave_agent.py:295]:
```suggestion
```

File: behave_reportportal/behave_agent.py [hunk 6]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+            return
```
Comment (iivanou) [inline behave_reportportal/behave_agent.py:296]:
```suggestion
```

File: behave_reportportal/behave_agent.py [hunk 7]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+                    level
+                    for level in context._stack
+                    if level.get("@layer") == "feature"
+                ]
+            ),
+            None,
+        )
+        if not layer:
+            return
+        if not layer.get("@cleanups"):
```
Comment (iivanou) [inline behave_reportportal/behave_agent.py]:
```suggestion
```

File: behave_reportportal/behave_agent.py [hunk 8]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+            return
```
Comment (iivanou) [inline behave_reportportal/behave_agent.py]:
```suggestion
```

File: behave_reportportal/behave_agent.py [hunk 9]
```diff
@@ -198,18 +254,121 @@ def _finish_step_scenario_based(self, step, **kwargs):
...
+        for cleanup in layer.get("@cleanups"):
```
Comment (iivanou) [inline behave_reportportal/behave_agent.py]:
```suggestion
        for cleanup in layer.get("@cleanups", []):
```

File: tox.ini [hunk 10]
```diff
@@ -6,3 +6,9 @@ multi_line_output=3
 force_grid_wrap=0
 combine_as_imports=True
 lines_after_imports=2
+
+[flake8]
+ignore = E203,W503
```
Comment (iivanou) [inline tox.ini:11]: Why can't we write a code according to PEP rather than ignore rules?

File: tests/units/test_rp_agent.py [hunk 11]
```diff
@@ -427,3 +538,298 @@ def test_analytics(mock_send_event, config):
...
+            ),
+            mock.call(
+                name="Execution of 'cleanup_func2' cleanup function",
+                start_time=123,
+                item_type="AFTER_TEST",
+                parent_item_id="scenario_id",
+            ),
+        ]
+    )
+    assert mock_rps.finish_test_item.call_count == 2
```
Comment (iivanou) [inline tests/units/test_rp_agent.py]: Make it softly.

File: behave_reportportal/behave_agent.py (hunk 1 above)
Comment (amartiushkov-ms) [reply behave_reportportal/behave_agent.py]: updated, thanks

File: behave_reportportal/behave_agent.py (hunk 2 above)
Comment (amartiushkov-ms) [reply behave_reportportal/behave_agent.py]: done

File: README.rst (hunk 3 above)
Comment (amartiushkov-ms) [reply README.rst]: The idea is to provide ability not to mark skipped tests as Issue in RP. they are not marked with some 'issue' attribute in code

File: behave_reportportal/behave_agent.py (hunk 4 above)
Comment (amartiushkov-ms) [reply behave_reportportal/behave_agent.py]: updated, thanks

File: behave_reportportal/behave_agent.py (hunk 5 above)
Comment (amartiushkov-ms) [reply behave_reportportal/behave_agent.py:295]: Why? tags is None when it is empty

File: behave_reportportal/behave_agent.py (hunk 9 above)
Comment (amartiushkov-ms) [reply behave_reportportal/behave_agent.py]: updated

File: tox.ini (hunk 10 above)
Comment (amartiushkov-ms) [reply tox.ini:11]: These changes were made automatically by black tool. According to [this](https://github.com/psf/black/issues/315) conversation I decided to ignore it.

File: tests/units/test_rp_agent.py (hunk 11 above)
Comment (amartiushkov-ms) [reply tests/units/test_rp_agent.py]: Why?

Comment (iivanou) [reply tests/units/test_rp_agent.py]: In case of raise condition, the next checks won't be run.

**Review by HardNorth — APPROVED** [review]

File: tests/units/test_rp_agent.py (hunk 11 above)
Comment (amartiushkov-ms) [reply tests/units/test_rp_agent.py]: It was the last verification in scope of test. Anyway mock assert method is used in this test and it's not soft

Comment (iivanou) [reply tests/units/test_rp_agent.py]: 1. Mock assert can be used with delayed_assert;
2. I see assert_has_calls call after this assert.
image

File: README.rst (hunk 3 above)
Comment (iivanou) [reply README.rst]: It's not about the content but about spelling.

File: tox.ini (hunk 10 above)
Comment (iivanou) [reply tox.ini:11]: Got it.

File: README.rst (hunk 3 above)
Comment (amartiushkov-ms) [reply README.rst]: From my point of view it changes the meaning of the comment. I can suggest comment from client "option to mark skipped tests as not 'To Investigate' items on Server side."

File: tests/units/test_rp_agent.py (hunk 11 above)
Comment (amartiushkov-ms) [reply tests/units/test_rp_agent.py]: This part of code was updated. It's last assert for test

File: README.rst (hunk 3 above)
Comment (amartiushkov-ms) [reply README.rst]: done

**Review by iivanou — APPROVED** [review]

**Pull request merged**